mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala

```diff
@@ -1041,13 +1041,13 @@ object ALS extends DefaultParamsReadable[ALS] with Logging {
       .persist(finalRDDStorageLevel)
     if (finalRDDStorageLevel != StorageLevel.NONE) {
       userIdAndFactors.count()
-      itemIdAndFactors.count()
-      itemFactors.unpersist()
       userInBlocks.unpersist()
       userOutBlocks.unpersist()
-      itemInBlocks.unpersist()
       itemOutBlocks.unpersist()
       blockRatings.unpersist()
+      itemIdAndFactors.count()
```

Comment:
I don't see the problem here immediately. Doesn't this depend on some of the cached user RDDs?

Reply:
itemIdAndFactors only directly depends on itemFactors and itemInBlocks.

Reply:
itemIdAndFactors only directly depends on itemFactors and itemInBlocks.

Reply:
Yes, but itemFactors depends on the previous user RDDs, including userOutBlocks. I think this is done on purpose.

Reply:
Yep, that's my point.